content/posts/lets-code-contact-form-in-phoenix-part-two.md

```diff
@@ -0,0 +1,264 @@
+coverCredit: "Photo by [Markus Winkler](https://unsplash.com/@markuswinkler?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText) on [Unsplash](https://unsplash.com/?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText)"
+draft: true
+comments: false
+tags: ["elixir", "phoenix"]
+---
+
+In this _Let's Code_ series, I will be walking through how I added a simple contact form to a [Phoenix](https://www.phoenixframework.org/) 1.4 web application. I will try to make this as general purpose as possible so that it's easier to follow _and_ apply to your Phoenix applications. I encourage you to follow along, either in an existing Phoenix application or [a new one](https://hexdocs.pm/phoenix/up_and_running.html#content).
+
+[In Part One]({{< ref "lets-code-contact-form-in-phoenix-part-one" >}}) of this series, I started with a `Message` module to serve as the data object for the contact form. The contact form should take the minimal number of fields from the user (email, subject, and body) and send that information to a support email address. In this post, I will be walking through creating a module to send the feedback. Let's get started.
+
+## The Support Interface
+
+For the purposes of a contact form, I know I will need to allow an end user to modify and submit a message to the support team. With that, I can determine two functions: `change_message` and `send_message`. The `change_message` function will be used to represent and validate the data on the form while the `send_message` function will be used to send an email to the support team with the contents of the message. But where do these functions live?
+
+As mentioned in the [previous post]({{< ref "lets-code-contact-form-in-phoenix-part-one" >}}), the Phoenix guides encourage defining [context modules](https://hexdocs.pm/phoenix/contexts.html#content) or public interfaces as a boundary around a specific domain of an application. The `Support` context will be the boundary and public interface for the Support domain in this application. Functions on this public interface can be called from other interfaces such as a Phoenix controller or an IEx session.
+
+In true Test-Driven Development (TDD) fashion, I can begin with a test file for this new module and functions:
+
+```elixir
+# test/my_app/support_test.exs
+
+defmodule MyApp.SupportTest do
+  # The `DataCase` module ships with a standard Phoenix app.
+  # See the docs for more info:
+  # https://hexdocs.pm/phoenix/testing_contexts.html#the-datacase
+  #
+  # The `async: true` option allows the tests in this file to run concurrently
+  # with other tests in the application. Tests in this file still run serially.
+  use MyApp.DataCase, async: true
+
+  alias MyApp.Support
+
+  test "changing a message" do
+    # `Support.change_message/1` is expected to return a Changeset for a Message.
+    assert %Ecto.Changeset{data: %Support.Message{}} = Support.change_message()
```

Comment:
I wonder if engineers coming from a Ruby world would find this syntax surprising (using `=` instead of `==`, or an `assert_equal` equivalent). Is it worth calling out that this assertion is using pattern matching, and therefore we can be sure that the `change_message` function returns the exact type that we're expecting?

Reply:
I think it's worth drawing attention to much in the same way that I now draw attention to the use of a guard clause 👍